### PR TITLE
Create jsonapi_tables submodule to allow accesses watchdog(and other tables) via JSON API

### PR DESCRIPTION
**Pull request by tedbow**
This is continuation of https://github.com/jsdrupal/drupal-admin-ui/pull/59

I no longer have access to  push to that branch.  The branch on my repo and this the other PR except latest merge from master.

I would consider this done. So can either be merged here or the other branch on jsdrupal updated and merged.

### CODE REVIEW

File: admin_ui_support/modules/jsonapi_tables/src/Entity/ReadOnlyTableEntityBase.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    parent::postLoad($storage, $entities);
+    $schema_fields = [];
+    /** @var \Drupal\Core\Entity\EntityInterface $entity */
+    foreach ($entities as $entity) {
+      if (!$schema_fields) {
+        $schema_fields = self::getEntityTypeSchemaFields($entity->getEntityType());
+      }
+      foreach ($entity as $field_name => $field) {
+        if (isset($schema_fields[$field_name])) {
+          if ($schema_fields[$field_name]['type'] === 'blob') {
```
Comment (dawehner) [inline admin_ui_support/modules/jsonapi_tables/src/Entity/ReadOnlyTableEntityBase.php]: Looking at ```\Drupal\menu_link\Plugin\Field\FieldType\MenuLinkItem::schema``` I'd say no. Could we add some indicator to base fields instead?

File: admin_ui_support/modules/jsonapi_tables/src/Entity/ReadOnlyTableEntityBase.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  /**
+   * Json serializes the value of blob field.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The entity.
+   * @param string $field_name
+   *   The blob field name.
+   */
+  protected static function jsonSerializeBlogField(EntityInterface $entity, $field_name) {
+    $value = $entity->get($field_name)[0]->getValue()['value'];
```
Comment (dawehner) [inline admin_ui_support/modules/jsonapi_tables/src/Entity/ReadOnlyTableEntityBase.php]: Is there a reason we don't just support all deltas?

File: admin_ui_support/modules/jsonapi_tables/src/Entity/WatchdogEntity.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ *     "storage" = "Drupal\jsonapi_tables\ReadOnlyTableEntityStorage",
+ *     "access" = "Drupal\jsonapi_tables\ReadOnlyAccessControlHandler",
+ *   },
+ *   base_table = "watchdog",
+ *   entity_keys = {
+ *     "id" = "wid",
+ *     "label" = "message",
+ *   },
+ *   links = {
+ *     "canonical" = "/{admin/reports/dblog/{watchdog_entity}",
```
Comment (dawehner) [inline admin_ui_support/modules/jsonapi_tables/src/Entity/WatchdogEntity.php]:
```+ *     "canonical" = "/{admin/reports/dblog/{watchdog_entity}",
```
I'm quite astonished that the leading ```/{``` actually works.

File: admin_ui_support/modules/jsonapi_tables/src/Entity/WatchdogEntity.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function postLoad(EntityStorageInterface $storage, array &$entities) {
+    parent::postLoad($storage, $entities);
+    foreach ($entities as &$entity) {
+      if (isset($entity->variables)) {
+        // Format the message of a watchdog entry.
+        $variables = $entity->variables[0]->getValue()['value'];
```
Comment (dawehner) [inline admin_ui_support/modules/jsonapi_tables/src/Entity/WatchdogEntity.php]: yeah as written above, what about using ```->setSetting('serialize', TRUE)``` for the base field

File: admin_ui_support/modules/jsonapi_tables/src/ReadOnlyTableEntityStorage.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function delete(array $entities) {
+    throw new \LogicException('Deleting not allow on read-only entity type');
+  }
+
+}
```
Comment (dawehner) [inline admin_ui_support/modules/jsonapi_tables/src/ReadOnlyTableEntityStorage.php]: Nitpick^10: Missing newline :)

File: admin_ui_support/modules/jsonapi_tables/src/Entity/ReadOnlyTableEntityBase.php (hunk 2 above)
Comment (tedbow) [reply admin_ui_support/modules/jsonapi_tables/src/Entity/ReadOnlyTableEntityBase.php]: We can't actually create a mutlivalue field when exposing a table like we are.

File: admin_ui_support/modules/jsonapi_tables/src/ReadOnlyTableEntityStorage.php (hunk 5 above)
Comment (tedbow) [reply admin_ui_support/modules/jsonapi_tables/src/ReadOnlyTableEntityStorage.php]: Fixed

File: admin_ui_support/modules/jsonapi_tables/src/Entity/ReadOnlyTableEntityBase.php (hunk 1 above)
Comment (tedbow) [reply admin_ui_support/modules/jsonapi_tables/src/Entity/ReadOnlyTableEntityBase.php]: Added a  `\Drupal\jsonapi_tables\EntityTypeTableSchemaTrait::isSerializedField` with todo to figure this out. 

for now if `serialize` is set it respects that. Then if it is a `blog` assumes it is serialized.

File: admin_ui_support/modules/jsonapi_tables/jsonapi_tables.info.yml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+name: 'JSONAPI Table'
```
[the remark this answers is not in the document]
Comment (dawehner) [reply admin_ui_support/modules/jsonapi_tables/jsonapi_tables.info.yml]: Well, it exposes simple tables using a custom entity type :)

**Pull request merged**